sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala

```diff
@@ -205,6 +224,108 @@ case class BroadcastNestedLoopJoinExec(
+
+  private def notInSubquerySingleColumnOptimizeEnabled: Boolean = {
+    if (SQLConf.get.notInSubquerySingleColumnOptimizeEnabled && right.output.length == 1) {
+      // BuildSide must be single column, condition must be the following pattern
+      // Or(EqualTo(a, b), IsNull(EqualTo(a, b)))
+      condition.get match {
+        case _ @ Or(
+            _ @ EqualTo(leftAttr: AttributeReference, rightAttr: AttributeReference),
+            _ @ IsNull(_ @ EqualTo(tmpLeft: AttributeReference, tmpRight: AttributeReference)))
+          if leftAttr.semanticEquals(tmpLeft) && rightAttr.semanticEquals(tmpRight) =>
+          notInSubquerySingleColumnOptimizeSetStreamedKey(leftAttr, rightAttr)
+          if (notInSubquerySingleColumnOptimizeStreamedKeyIndex != -1) {
+            true
+          } else {
+            logWarning(s"failed to find notInSubquerySingleColumnOptimizeStreamedKeyIndex," +
+              s" fallback to leftExistenceJoin.")
+            false
+          }
+        case _ => false
+      }
+    } else {
+      false
+    }
+  }
+
+  private def notInSubquerySingleColumnOptimizeBuildParams(
+      buildRows: Array[InternalRow]): NotInSubquerySingleColumnOptimizeParams = {
+    NotInSubquerySingleColumnOptimizeParams(
+      HashedRelation(buildRows.iterator,
+        BindReferences.bindReferences[Expression](
+          Seq(right.output.head), AttributeSeq(right.output)),
+        buildRows.length),
+      buildRows.exists(row => row.isNullAt(0)),
+      buildRows.isEmpty)
+  }
+
+  private def notInSubquerySingleColumnOptimizeSetStreamedKey(
+      leftAttr: AttributeReference,
+      rightAttr: AttributeReference): Unit = {
+    val leftNotInKeyFound = left.output.exists(_.semanticEquals(leftAttr))
+    if (leftNotInKeyFound) {
+      notInSubquerySingleColumnOptimizeStreamedKey = leftAttr
+      notInSubquerySingleColumnOptimizeStreamedKeyIndex =
+        AttributeSeq(left.output).indexOf(leftAttr.exprId)
+    } else {
+      notInSubquerySingleColumnOptimizeStreamedKey = rightAttr
+      notInSubquerySingleColumnOptimizeStreamedKeyIndex =
+        AttributeSeq(left.output).indexOf(rightAttr.exprId)
+    }
+  }
+
+  /**
+   * Optimized version implementation for LeftAnti
+   * which converted from single column NotInSubquery
+   * @param relation
+   * @return
+   */
+  private def notInSubquerySingleColumnOptimize(
+      relation: Broadcast[Array[InternalRow]]): RDD[InternalRow] = {
+    assert(buildSide == BuildRight)
+    val buildRows = relation.value
+    if (buildRows.length > SQLConf.get.notInSubquerySingleColumnOptimizeRowCountThreshold) {
+      logWarning(s"BuildSide row count ${buildRows.length} exceeded " +
+        s"notInSubquerySingleColumnOptimizeRowCountThreshold ${SQLConf.get
+          .notInSubquerySingleColumnOptimizeRowCountThreshold}, fallback to leftExistenceJoin.")
+      leftExistenceJoin(relation, false)
+    } else {
+      val params = notInSubquerySingleColumnOptimizeBuildParams(buildRows)
```

Comment:
This Null aware single column hash join is subtle and could use some comments or perhaps a reference. Perhaps you could link to section 6.1 of that paper above ?

Reply:
done. with TODO and comment left.